Consul API Gateway v.0.2.0 upgrade path documentation (#12812)
hashicorp/consul: website/content/docs/api-gateway/upgrade-specific-versions.mdx

---
layout: docs
page_title: Upgrading Specific Versions
description: >-
  Consul api gateway versions v0.2.x introduced a change to namespace backend ref
  enforcement that will potentially break any routes created with a cross namespace in
  v0.1.0. This documents the flow for finding and remediating any potentially
  broken routes
---

# Upgrading Specfic Versions

## Consul API Gateway v0.1.0

This guide explains how to best upgrade a consul-api-gateway deployment that is currently
running v0.1.0 to all future versions. v0.2.0 introduced a breaking change that
causes routes defined with a BackendRef defined in a different namespace to
require a [ReferencePolicy](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.ReferencePolicy)
that explicitly allows traffic from the routes namespace to the backendrefs namespace.

## Requirements

- The consul-api-gateway should be running version v0.1.0, if the version is different, follow the normal upgrade path

> - [jq](https://stedolan.github.io/jq/download/) is installed on the users CLI

Reviewer: Let's treat it as optional. First explain what needs to be done (identify which gateways and routes cross namespaces, and create reference policies for them). Then say here is an optional way to make it easier.

> - [jq](https://stedolan.github.io/jq/download/) is installed on the users CLI

Author: Alright I marked it as optional and wrote a section on how to pull the routes manually in addition to the jq command

Note, as of writing, this is the only upgrade path documented as v0.1.0 is the only currently released version. Normal upgrade path will be documented
in future releases

## Assumptions

This guides makes the following assumptions:

- You have the ability to run Kubectl cli commands
- Your kubectl config is already configured to point at the cluster with the installation you are upgrading
- You have HTTPRoute.read, TCPRoute.read and ReferencePolicy.create rights on your kubernetes cluster

## Procedures

**1.** Optional, double check the current version of the consul-api-gateway with the following command

```

> kubectl get deployment --namespace consul consul-api-gateway-controller -o json | jq '.spec.template.spec.containers[] | select(.name="api-gateway-controller") .image'

Reviewer 2: ```suggestion
kubectl get deployment --namespace consul consul-api-gateway-controller --output=jsonpath= "{@.spec.template.spec.containers[?(@.name=='api-gateway-controller')].image}"
```

```

If the output looks as follows

```
"hashicorp/consul-api-gateway:0.1.0"
```

 continue following this upgrade path.

**2.** Run the following command to retrieve all routes defined with a cross namespace:

```

> kubectl get HttpRoute -o json -A | jq -r '.items[] | {name: .metadata.name, namespace: .metadata.namespace, kind: .kind, crossNamespaceBackendReferences: ( .metadata.namespace as $parentnamespace | .spec.rules[] .backendRefs[] | select(.namespace != null and .namespace != $parentnamespace )  )}  ' && kubectl get TcpRoute -o json -A | jq -r '.items[] | {name: .metadata.name, namespace: .metadata.namespace, kind: .kind, crossNamespaceBackendReferences: ( .metadata.namespace as $parentnamespace | .spec.rules[] .backendRefs[] | select(.namespace != null and .namespace != $parentnamespace )  )}  '
> 

Reviewer 2: I'm not entirely sure without testing this myself, but I _think_ you can simplify this command by performing a simultaneous get of both `TcpRoute` and `HttpRoute`, then parsing the items from a single array.

```suggestion
kubectl get HttpRoute,TcpRoute -o json -A | jq -r '.items[] | {name: .metadata.name, namespace: .metadata.namespace, kind: .kind, crossNamespaceBackendReferences: ( .metadata.namespace as $parentnamespace | .spec.rules[] .backendRefs[] | select(.namespace != null and .namespace != $parentnamespace )  )} '
```

Author: That worked, good call, thank you!

```

Note, the above command will retrieve all HTTPRoutes and TCPRoutes. TLSRoutes and UDPRoutes are not supported in v0.1.0

If you have any breaking namespaces, your output will something like as follows. If your output is empty, you have no breaking routes and you can skip to the last step.

```
{
  "name": "example-breaking-http-route",
  "namespace": "example-namespace",
  "kind": "HTTPRoute",
  "crossNamespaceBackendReferences": {
    "group": "",
    "kind": "Service",
    "name": "web-backend",
    "namespace": "gateway-namespace",
    "port": 8080,
    "weight": 1
  }
}
{
  "name": "example-breaking-tcp-route",
  "namespace": "a-different-namespace",
  "kind": "TCPRoute",
  "crossNamespaceBackendReferences": {
    "group": "",
    "kind": "Service",
    "name": "web-backend",
    "namespace": "gateway-namespace",
    "port": 8080,
    "weight": 1
  }
}
```

**3.** Create a ReferencePolicy to allow cross namespace traffic.

> 

Reviewer: We need to explain that a ReferencePolicy needs to be created for each pair of route/service namespaces.

Using the above output as a guide, create a [ReferencePolicy](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.ReferencePolicy) to allow cross namespace traffic. If you've already created a ReferencePolicy you can skip this step.

> 

Reviewer: ```suggestion
<!---
TODO: add link to our docs on Cross Namespace Reference Policies, once we have written then, and tell the user to see them for more details on how to create these policies.
--->
```

For example, the above output would require a reference policy that looks follows.

Note, the ReferencePolicy should be created in the same namespace that its allowing traffic to

<CodeBlockConfig filename="referencepolicy.yaml">

apiVersion: gateway.networking.k8s.io/v1alpha2
kind: ReferencePolicy
metadata:
  name: reference-policy
  namespace: gateway-namespace
spec:
  from:
    - group: gateway.networking.k8s.io
      kind: HTTPRoute
      namespace: example-namespace
    - group: gateway.networking.k8s.io
      kind: TCPRoute
      namespace: a-different-namespace
  to:

>     - group: ""

Author: I actually get the following error when I try remove group and specify namespace in the to field in the ReferencePolicy spec

```
error: error validating "test.yaml": error validating data: [ValidationError(ReferencePolicy.spec.to[0]): unknown field "namespace" in io.k8s.networking.gateway.v1alpha2.ReferencePolicy.spec.to, ValidationError(ReferencePolicy.spec.to[0]): missing required field "group" in io.k8s.networking.gateway.v1alpha2.ReferencePolicy.spec.to]; if you choose to ignore these errors, turn validation off with --validate=false
```

My assumption is that this is because the ReferencePolicy assumes that its directing traffic to the same namespace that the reference policy is in, but I could be wrong about that.

>     - group: ""

Reviewer 3: Namespace doesn't get specified for `to` since it's assumed to be the namespace that the ReferencePolicy is deployed into. You can just drop `group: ""` and should be good to go

>     - group: ""

Author: I get 

```error: error validating "test.yaml": error validating data: ValidationError(ReferencePolicy.spec.to[0]): missing required field "group" in io.k8s.networking.gateway.v1alpha2.ReferencePolicy.spec.to; if you choose to ignore these errors, turn validation off with --validate=false```

When I drop group

>     - group: ""

Reviewer 3: oh interesting! I don't guess I've tried installing with group deleted. This has implications for https://github.com/hashicorp/tutorials/pull/260 then

>     - group: ""

Reviewer 4: Just saw this, that's silly. Think we should raise this upstream. If there's default behavior for an empty string (i.e. `core`), this should just be an optional field in the spec.

      kind: Service
      name: web-backend
</CodeBlockConfig>

Edit your ReferencePolicy so all routes are allowed and save into a file called referencepolicy.yaml

Run the following command to apply it to your cluster

```
kubectl apply -f referencepolicy.yaml
```

**4.** Upgrade your deployment to v.0.2.0.

```
kubectl apply --kustomize="github.com/hashicorp/consul-api-gateway/config/crd?ref=0.2.0"

> ```

Reviewer: Briefly explain what this command is doing.

**5.** Helm upgrade consul

```
helm upgrade --values values.yaml --namespace consul --version <NEW_VERSION> hashicorp/consul <DEPLOYMENT_NAME>

> ```

Reviewer: Briefly explain what this command is doing.  We need to add a warning note that this will cause the API Gateway controller to shut down and restart with the new version.

**6.** Refresh gateway deployments

> Because [the gateway class is immutable](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io%2fv1alpha2.GatewayClass), once you've upgraded your CRD installation, in order to see the effects on prexisting gateways, you will need to delete and recreate any gateways.

Reviewer: I've got comments I need to add here but I'll do it tomorrow.

Reviewer: Most people won't have enough contextual knowledge to understand what it means when we say the gateway class is "immutable." Instead, we can say something like:

```suggestion
A requirement of the Kubernetes Gateway API specification is that the settings of the [Gateway Class](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io%2fv1alpha2.GatewayClass) are only applied to a gateway when the gateway is created. For that reason, in order to see the effects on preexisting gateways, once you've upgraded your CRD installation,  you will need to delete and recreate any gateways.
```

---
layout: docs
page_title: Upgrading Specific Versions
description: >-
  Consul api gateway versions v0.2.x introduced a change to namespace backend ref
  enforcement that will potentially break any routes created with a cross namespace in
  v0.1.0. This documents the flow for finding and remediating any potentially
  broken routes
---

# Upgrading Specfic Versions

## Consul API Gateway v0.1.0

> This guide explains how to best upgrade a consul-api-gateway deployment that is currently

Reviewer 5: +1 , also to the above description

> This guide explains how to best upgrade a consul-api-gateway deployment that is currently

Reviewer: The only times we should use consul-api-gateway is when it is in a code block or when the instructions need to refer to the container image by name.

You will need to create a reference policy that explicitly allows each cross namespace route to service pair to prevent the route from breaking. If you've already created a ReferencePolicy you can skip this step.
<!---
TODO: add link to our docs on Cross Namespace Reference Policies, once we have written then, and tell the user to see them for more details on how to create these policies.
--->
For example, the above output would require a reference policy that looks as follows.

Note, the ReferencePolicy should be created in the same namespace that it is allowing traffic to

<CodeBlockConfig filename="referencepolicy.yaml">

apiVersion: gateway.networking.k8s.io/v1alpha2
kind: ReferencePolicy
metadata:
  name: reference-policy
  namespace: gateway-namespace
spec:
  from:
    - group: gateway.networking.k8s.io
      kind: HTTPRoute
      namespace: example-namespace
    - group: gateway.networking.k8s.io
      kind: TCPRoute
      namespace: a-different-namespace
  to:
    - group: ""
      kind: Service
      name: web-backend
</CodeBlockConfig>

> Edit your ReferencePolicy so all routes are allowed and save into a file called referencepolicy.yaml

Reviewer 5: and -> , then

---
layout: docs
page_title: Upgrading Specific Versions
description: >-
  Consul api gateway versions v0.2.x introduced a change to namespace backend ref
  enforcement that will potentially break any routes created with a cross namespace in
  v0.1.0. This documents the flow for finding and remediating any potentially

>   broken routes

Reviewer 3: It probably makes sense to generalize this since we're going to be tracking future release requirements on this page as well

Reviewer 3: This is slightly different from the intro updated in [20f3c4c](https://github.com/hashicorp/consul/pull/12812/commits/20f3c4c105a45b0b7c8d9a79c622e2a1b4120146). I'm not even sure if this one is visible to the public or just to us

Reviewer 3: @sarahalsmiller This is slightly different from the intro updated in [20f3c4c](https://github.com/hashicorp/consul/pull/12812/commits/20f3c4c105a45b0b7c8d9a79c622e2a1b4120146). I'm not even sure if this one is visible to the public or just to us

This guide explains how to upgrade a Consul API Gateway deployment running a specific version
that requires additional steps from the standard upgrade path.

Note: As of writing, v0.1.0 is the only previous release. A standardized upgrade path will be documented with future releases.

> ## Consul API Gateway v0.1.0

Reviewer: ```suggestion
## Consul API Gateway v0.2.0
```
This should be the version being upgraded **to**

---
layout: docs
page_title: Upgrading Specific Versions
description: >-
  Consul api gateway versions v0.2.x introduced a change to namespace backend ref
  enforcement that will potentially break any routes created with a cross namespace in
  v0.1.0. This documents the flow for finding and remediating any potentially
  broken routes
---

# Upgrading Specific Versions

## Consul API Gateway v0.1.0

This guide explains how to best upgrade a Consul API Gateway deployment that is currently
running v0.1.0 to all future versions. v0.2.0 introduced a breaking change that
causes routes defined with a BackendRef defined in a different namespace to
require a [ReferencePolicy](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.ReferencePolicy)
that explicitly allows traffic from the route's namespace to the BackendRef's namespace.

> ## Requirements

Author: I had it originally structured as one page, but we moved it to this pattern. I'll update the headers though, that's a good call.

You will need to create a reference policy that explicitly allows each cross namespace route to service pair to prevent the route from breaking. If you've already created a ReferencePolicy you can skip this step.
<!---
TODO: add link to our docs on Cross Namespace Reference Policies, once we have written then, and tell the user to see them for more details on how to create these policies.
--->
For example, the above output would require a reference policy that looks as follows.

Note, the ReferencePolicy should be created in the same namespace that it is allowing traffic to

<CodeBlockConfig filename="referencepolicy.yaml">

```yaml
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: ReferencePolicy
metadata:
  name: reference-policy
  namespace: gateway-namespace
spec:
  from:
    - group: gateway.networking.k8s.io
      kind: HTTPRoute
      namespace: example-namespace
    - group: gateway.networking.k8s.io
      kind: TCPRoute
      namespace: a-different-namespace
  to:
    - group: ""
      kind: Service
      name: web-backend
</div>
```
</CodeBlockConfig>

> Edit your ReferencePolicy so all routes are allowed and then save into a file called referencepolicy.yaml

Reviewer 3: I think it would be helpful to note that each `ReferencePolicy` may only specify one `from` and one `to` ([docs](https://gateway-api.sigs.k8s.io/v1alpha2/api-types/referencepolicy/#api-design-decisions)), so the Operator may need to define more than one `ReferencePolicy`.

> Each ReferencePolicy only supports a single From and To section. Additional trust relationships must be modeled with additional ReferencePolicy resources.

Run the following command to apply it to your cluster

```shell-session

> $ kubectl apply -f referencepolicy.yaml

Reviewer 3: We generally want to use the longer version of flag names everywhere to help out anyone reading the docs that isn't super familiar with kubectl

```suggestion
$ kubectl apply --filename referencepolicy.yaml
```

```

> Repeat as needed until each of your cross namespace routes have a corresponding  reference policys.

Reviewer 3: ```suggestion
Repeat as needed until each of your cross-namespace routes have a corresponding ReferencePolicy.
```

**4.** Upgrade your deployment to v.0.2.0.

Install the updated CRDs into your cluster

``` shell-session
$ kubectl apply --kustomize="github.com/hashicorp/consul-api-gateway/config/crd?ref=0.2.0"
```

**5.** Helm upgrade consul

> Upgrade your consul installation using helm.

Reviewer 3: ```suggestion
Upgrade your Consul installation using Helm.
```

> ~> This will call the API Gateway controller to be shut down and restart with the new version

Reviewer 3: ```suggestion
~> This will cause the Consul API Gateway controller to be shut down and restart with the new version
```

```shell-session
$ helm upgrade --values values.yaml --namespace consul --version <NEW_VERSION> hashicorp/consul <DEPLOYMENT_NAME>
```

**6.** Refresh gateway deployments
A requirement of the Kubernetes Gateway API specification is that the settings of the [Gateway Class](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io%2fv1alpha2.GatewayClass) are only applied to a gateway when the gateway is created. For that reason, in order to see the effects on preexisting gateways, once you've upgraded your CRD installation,  you will need to delete and recreate any gateways.

> To accomplish this you can run the following two commands

Reviewer 3: ```suggestion
To accomplish this you can run the following two commands:
```

~> This will delete and recreate your gateway.

```shell-session
$ kubectl delete -f <path_to_gateway_config.yaml>
$ kubectl create -f <path_to_gateway_config.yaml>
```

<!---
  remove this warning once updating a gateway triggers reconciliation on child routes
--->

> ~> Optionally you might want to delete and recreate your routes following the same pattern, as when you update your

Reviewer 3: ```suggestion
~> Optionally, you might want to delete and recreate your routes following the same pattern, as when you update your
```

> gateway, it might take a long time for it's attached routes to reconcile and start reporting bind errors.

Author: I don't think we're expecting errors, but if someone missed a referencepolicy its possible one would occur, right?

## Consul API Gateway v0.1.0

Consul API Gateway v0.2.0 introduced a breaking change that causes routes with a BackendRef defined in a different namespace to
require a [ReferencePolicy](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.ReferencePolicy)
that explicitly allows traffic from the route's namespace to the BackendRef's namespace. This guide explains how to find all routes
that require a ReferencePolicy and create a matching ReferencePolicy.

### Requirements

- The consul-api-gateway should be running version v0.1.0. If the version is different, follow the normal upgrade path.
- **Optional** [jq](https://stedolan.github.io/jq/download/) is installed on the users CLI

### Pre-requisites

In order to follow this guide, the following conditions must be true:

- You have the ability to run Kubectl CLI commands
- Your kubectl config is already configured to point at the cluster with the installation you are upgrading
- You have HTTPRoute.read, TCPRoute.read and ReferencePolicy.create rights on your Kubernetes cluster

### Procedures

> **1.** Optional, double check the current version of the consul-api-gateway with the following command:

Reviewer 3: ```suggestion
**1.** Double check the current version of the consul-api-gateway-controller Deployment with the following command:
```

If we think it's worth including, I wouldn't mark as optional. I think people will naturally skip it if they have full faith in themselves.

```shell-session
$ kubectl get deployment --namespace consul consul-api-gateway-controller --output=jsonpath= "{@.spec.template.spec.containers[?(@.name=='api-gateway-controller')].image}"
```

If the output looks as follows:

```log
"hashicorp/consul-api-gateway:0.1.0"
```

 Continue following this upgrade path.

> **2.** Retrieve all routes that have a backend in a different namespace:

Reviewer 3: ```suggestion
**2.** Retrieve all routes that have a backend in a different namespace
```

There are two ways to retrieve cross-namespace routes:

1. Method 1 is more manual but requires no additional dependencies
1. Method 2 is faster but requires that [jq](https://stedolan.github.io/jq/) be installed

##### Method 1: Manual

Get all HTTPRoutes and TCPRoutes across all namespaces with the following command:

```shell-session
$ kubectl get HTTPRoute,TCPRoute -o json -A
```

If you have any active HTTPRoutes or TCPRoutes, you should receive output that looks as follows. Note that the output has been truncated to show only relevant fields.

> Note, the above command will retrieve all HTTPRoutes and TCPRoutes. TLSRoutes and UDPRoutes are not supported in v0.1.0.

Reviewer 3: ```suggestion
Note that the above command will retrieve only HTTPRoutes and TCPRoutes. TLSRoutes and UDPRoutes are not supported in v0.1.0.
```

```yaml
...

apiVersion: v1
items:
- apiVersion: gateway.networking.k8s.io/v1alpha2
  kind: HTTPRoute
  metadata:
    name: example-http-route,
    namespace: example-namespace,
    ...
  spec:
    parentRefs:
    - group: gateway.networking.k8s.io
      kind: Gateway
      name: gateway
      namespace: gw-ns
    rules:
    - backendRefs:
      - group: ""
        kind: Service
        name: web-backend
        namespace: gateway-namespace
        ...
      ...
- apiVersion: gateway.networking.k8s.io/v1alpha2
  kind: TCPRoute
  metadata:
    name: example-tcp-route,
    namespace: a-different-namespace,
    ...
  spec:
    parentRefs:
    - group: gateway.networking.k8s.io
      kind: Gateway
      name: gateway
      namespace: gateway-namespace
    rules:
    - backendRefs:
      - group: ""
        kind: Service
        name: web-backend
        namespace: gateway-namespace
        ...
...

```

For each of the above defined routes, you will need to inspect each of the backendRefs.

> If a backendRef has no namespace field defined or the namespace matches the namespace of the route itself, that backendRef requires no further action.

Reviewer 3: I'm curious in general if we should be using inline code blocks (so `backendRef` instead of backendRef) where we mention things like backendRef, ReferencePolicy, Deployment, Service, HTTPRoute, TCPRoute, etc. 

I didn't see any clear guidance on this in the [Markdown](https://github.com/hashicorp/engineering-docs/blob/main/writing/markdown.md#inline-styles) or [Writing](https://github.com/hashicorp/engineering-docs/blob/main/writing/style-guide.md#text-formatting) style guides.

Could you speak to this @Jeff-Apple?

Reviewer: @nathancoleman , When referring to a config parameter/object/etc. that the user is dealing with, we usually use prefer inline code blocks. Here are two examples of when I would and wouldn't use inline code blocks:

"When you configure a `HTTPRoute`, you will..."
"When you configure a route, you will..."

For "dictionary" words such as gateway, that are also config parameters, it depends on if the word is being used in a generic way or referring to the config parameter. Here is how I would write the sentence from line 129:

"If a `backendRef` does not have a `namespace` field defined or the namespace matches the namespace of the route itself, that `backendRef` requires no further action."

Reviewer 6: Once we're able to manage docs in our own repo/versioning it may be nice to add a lint task to check for improper capitalization of these terms.

$ kubectl get HTTPRoute,TCPRoute -o json -A | jq -r '.items[] | {name: .metadata.name, namespace: .metadata.namespace, kind: .kind, crossNamespaceBackendReferences: ( .metadata.namespace as $parentnamespace | .spec.rules[] .backendRefs[] | select(.namespace != null and .namespace != $parentnamespace )  )} '
```

Note, the above command will retrieve all HTTPRoutes and TCPRoutes. TLSRoutes and UDPRoutes are not supported in v0.1.0.

If your output is empty, you can skip to the last step; otherwise, you have existing routes that require a new ReferencePolicy. In this case, your output will appear similar to the following:

```log
{
  "name": "example-http-route",
  "namespace": "example-namespace",
  "kind": "HTTPRoute",
  "crossNamespaceBackendReferences": {
    "group": "",
    "kind": "Service",
    "name": "web-backend",
    "namespace": "gateway-namespace",
    "port": 8080,
    "weight": 1
  }
}
{
  "name": "example-tcp-route",
  "namespace": "a-different-namespace",
  "kind": "TCPRoute",
  "crossNamespaceBackendReferences": {
    "group": "",
    "kind": "Service",
    "name": "web-backend",
    "namespace": "gateway-namespace",
    "port": 8080,
    "weight": 1
  }
}
```

> **3.** Create a ReferencePolicy to allow cross namespace traffic for each route service pair.

Reviewer 3: ```suggestion
**3.** Create a ReferencePolicy to allow cross namespace traffic for each route service pair
```
